File: parser.go
package main

type Parser struct {
	tokenizer *Tokenizer
}

func NewParser(tokenizer *Tokenizer) *Parser {
	return &Parser{
		tokenizer: tokenizer,
	}
}

func (p *Parser) Parse() (program *Program, err error) {
	defer func() {
		err = toErr(recover())
	}()

	program = &Program{}
	for {
		stmt := p.parseGlobalStatement()
		if stmt == nil {
			break
		}
		program.stmts = append(program.stmts, stmt)
	}
	tk := p.tokenizer.Next()
	if tk.typ != ttEOF {
		panic("unexpected statement")
	}

	return program, nil
}

func (p *Parser) expect(tt TokenType) Token {
	token := p.tokenizer.Next()
	if token.typ != tt {
		panicf("unexpected token: %v", token)
	}
	return token
}

func (p *Parser) match(tts ...TokenType) (Token, bool) {
	tk := p.tokenizer.Next()
	for _, tt := range tts {
		if tk.typ == tt {
			return tk, true
		}
	}
	p.tokenizer.Undo(tk)
	return Token{}, false
}

func (p *Parser) next() Token {
	return p.tokenizer.Next()
}

func (p *Parser) skip(tt TokenType) bool {
	if p.tokenizer.Peek().typ == tt {
		p.tokenizer.Next()
		return true
	}
	return false
}

func (p *Parser) peek() Token {
	return p.tokenizer.Peek()
}

func (p *Parser) parseGlobalStatement() Statement {
	return p.parseStatement(true)
}

func (p *Parser) parseStatement(global bool) Statement {
	tk := p.tokenizer.Peek()

	switch tk.typ {
	case ttLet:
		return p.parseVariableStatement()
	case ttFunction:
		stmt := p.parseFunctionStatement()
		fn := stmt.(*FunctionStatement)
		if fn.name == "" {
			panic("function statement must have function name")
		}
		return stmt
	}

	if global {
		return nil
	}

	switch tk.typ {
	case ttReturn:
		return p.parseReturnStatement()
	case ttLeftBrace:
		// Notice: block statement skips parsing {} as object literal.
		return p.parseBlockStatement()
	case ttWhile:
		return p.parseWhileStatement()
	case ttBreak:
		return p.parseBreakStatement()
	case ttIf:
		return p.parseIfStatement()
	}

	if stmt := p.parseExpressionStatement(); stmt != nil {
		return stmt
	}
	if stmt := p.parseAssignmentStatement(); stmt != nil {
		return stmt
	}

	return nil
}

func (p *Parser) parseExpression() Expression {
	return p.parseEqualityExpression()
}

func (p *Parser) parseVariableStatement() Statement {
	var v VariableStatement
	p.expect(ttLet)
	v.Name = p.expect(ttIdentifier).str
	if p.tokenizer.Peek().typ == ttAssign {
		p.tokenizer.Next()
		v.Expr = p.parseExpression()
	}
	p.expect(ttSemicolon)
	return &v
}

func (p *Parser) parseAssignmentStatement() (stmt Statement) {
	p.tokenizer.PushFrame()
	defer func() {
		p.tokenizer.PopFrame(stmt == nil)
	}()

	var as VariableAssignmentStatement
	name := p.tokenizer.Next()
	if name.typ != ttIdentifier &&
		name.typ != ttBoolean && // these two are predeclared constants
		name.typ != ttNil {
		p.tokenizer.Undo(name)
		return nil
	}
	as.Name = name.str

	assign := p.tokenizer.Next()
	if assign.typ != ttAssign {
		p.tokenizer.Undo(assign)
		p.tokenizer.Undo(name)
		return nil
	}

	as.Expr = p.parseExpression()

	semi := p.tokenizer.Next()
	if semi.typ != ttSemicolon {
		p.tokenizer.Undo(semi)
		return nil
	}

	if name.typ == ttBoolean || name.typ == ttNil {
		panicf("predeclared constants cannot be assigned: %v", name)
	}

	return &as
}

func (p *Parser) parseFunctionStatement() Statement {
	var fn FunctionStatement
	p.expect(ttFunction)
	expr := p.parseFunctionExpression().(*FunctionExpression)
	fn.name = expr.name
	fn.expr = expr
	return &fn
}

func (p *Parser) parseReturnStatement() Statement {
	p.expect(ttReturn)
	expr := p.parseExpression()
	p.expect(ttSemicolon)
	return &ReturnStatement{
		expr: expr,
	}
}

func (p *Parser) parseExpressionStatement() (stmt Statement) {
	p.tokenizer.PushFrame()
	defer func() {
		p.tokenizer.PopFrame(stmt == nil)
	}()

	expr := p.parseExpression()
	if expr == nil {
		return nil
	}
	stmt = &ExpressionStatement{
		expr: expr,
	}
	semi := p.tokenizer.Next()
	if semi.typ == ttSemicolon {
		return stmt
	}
	p.tokenizer.Undo(semi)
	return nil
}

func (p *Parser) parseBlockStatement() (stmt Statement) {
	if p.tokenizer.Peek().typ != ttLeftBrace {
		return nil
	}

	block := &BlockStatement{}
	p.expect(ttLeftBrace)
	for {
		stmt := p.parseStatement(false)
		if stmt == nil {
			break
		}
		block.stmts = append(block.stmts, stmt)
	}
	p.expect(ttRightBrace)

	return block
}

func (p *Parser) parseWhileStatement() Statement {
	p.expect(ttWhile)
	expr := p.parseExpression()
	stmt := p.parseBlockStatement()
	return &WhileStatement{
		expr:  expr,
		block: stmt.(*BlockStatement),
	}
}

func (p *Parser) parseBreakStatement() Statement {
	p.expect(ttBreak)
	p.expect(ttSemicolon)
	return &BreakStatement{}
}

func (p *Parser) parseIfStatement() Statement {
	p.expect(ttIf)
	expr := p.parseExpression()
	ifBlock := p.parseBlockStatement()
	var elseBlock Statement
	switch p.tokenizer.Peek().typ {
	case ttElse:
		p.expect(ttElse)
		switch p.tokenizer.Peek().typ {
		case ttIf:
			elseBlock = p.parseIfStatement()
		case ttLeftBrace:
			elseBlock = p.parseBlockStatement()
		default:
			panic("else expect if or block to follow")
		}
	}
	return &IfStatement{
		cond:      expr,
		ifBlock:   ifBlock.(*BlockStatement),
		elseBlock: elseBlock,
	}
}

func (p *Parser) parseEqualityExpression() Expression {
	left := p.parseComparisonExpression()
	for {
		if op, ok := p.match(ttEqual, ttNotEqual); ok {
			right := p.parseComparisonExpression()
			left = NewBinaryExpression(left, op.typ, right)
		} else {
			break
		}
	}
	return left
}

func (p *Parser) parseComparisonExpression() Expression {
	left := p.parseAdditionExpression()
	for {
		if op, ok := p.match(ttGreaterThan, ttGreaterThanOrEqual, ttLessThan, ttLessThanOrEqual); ok {
			right := p.parseAdditionExpression()
			left = NewBinaryExpression(left, op.typ, right)
		} else {
			break
		}
	}
	return left
}

func (p *Parser) parseAdditionExpression() Expression {
	left := p.parseMultiplicationExpression()
	for {
		if op, ok := p.match(ttAddition, ttSubstraction); ok {
			right := p.parseMultiplicationExpression()
			left = NewBinaryExpression(left, op.typ, right)
		} else {
			break
		}
	}
	return left
}

func (p *Parser) parseMultiplicationExpression() Expression {
	left := p.parseUnaryExpression()
	for {
		if op, ok := p.match(ttMultiply, ttDivision); ok {
			right := p.parseUnaryExpression()
			left = NewBinaryExpression(left, op.typ, right)
		} else {
			break
		}
	}
	return left
}

func (p *Parser) parseUnaryExpression() Expression {
	if op, ok := p.match(ttNot, ttSubstraction); ok {
		right := p.parseUnaryExpression()
		return NewUnaryExpression(op.typ, right)
	}
	return p.parsePrimaryExpression()
}

func (p *Parser) parsePrimaryExpression() Expression {
	var expr Expression
	next := p.tokenizer.Next()

	switch next.typ {
	case ttNil:
		expr = ValueFromNil()
	case ttBoolean:
		expr = ValueFromBoolean(next.str == "true")
	case ttNumber:
		expr = ValueFromNumber(next.num)
	case ttString:
		expr = ValueFromString(next.str)
	case ttLeftParen:
		expr = p.parseExpression()
		p.expect(ttRightParen)
	case ttIdentifier:
		expr = ValueFromVariable(next.str)
	case ttFunction:
		expr = p.parseFunctionExpression()
	case ttLeftBrace:
		p.tokenizer.Undo(next)
		expr = p.parseObjectExpression()
	default:
		p.tokenizer.Undo(next)
		return nil
	}

	for {
		call := p.parseCallExpression()
		if call != nil {
			callExpr := call.(*CallExpression)
			callExpr.Callable = expr
			expr = call
			continue
		}
		break
	}

	return expr
}

func (p *Parser) parseCallExpression() Expression {
	if paren := p.tokenizer.Next(); paren.typ != ttLeftParen {
		p.tokenizer.Undo(paren)
		return nil
	}

	call := CallExpression{}
	call.Args = &Arguments{}

	for {
		arg := p.parseExpression()
		if arg == nil {
			break
		}
		call.Args.PutArgument(arg)
		if comma := p.tokenizer.Next(); comma.typ != ttComma {
			p.tokenizer.Undo(comma)
			break
		}
	}

	p.expect(ttRightParen)

	return &call
}

func (p *Parser) parseFunctionExpression() Expression {
	var name string
	var block *BlockStatement
	params := &Parameters{}

	if p.tokenizer.Peek().typ == ttIdentifier {
		name = p.tokenizer.Next().str
	}

	p.expect(ttLeftParen)
	for {
		tk := p.tokenizer.Next()
		if tk.typ == ttIdentifier {
			params.PutParam(tk.str)
		} else if tk.typ == ttComma {
			continue
		} else if tk.typ == ttRightParen {
			p.tokenizer.Undo(tk)
			break
		}
	}
	p.expect(ttRightParen)

	if p.tokenizer.Peek().typ != ttLeftBrace {
		panic("function needs a body")
	}

	block = p.parseBlockStatement().(*BlockStatement)

	return &FunctionExpression{
		name:   name,
		params: params,
		block:  block,
	}
}

func (p *Parser) parseObjectExpression() Expression {
	objexpr := NewObjectExpression()

	var key string
	var expr Expression

	p.expect(ttLeftBrace)

	for {
		token := p.next()

		switch token.typ {
		case ttString:
			key = token.str
		case ttIdentifier:
			key = token.str
		default:
			panic("unsupported key type")
		}

		p.expect(ttColon)

		expr = p.parseExpression()
		objexpr.props[key] = expr

		p.skip(ttComma)

		if p.skip(ttRightBrace) {
			break
		}
	}

	return objexpr
}

File: cppsrc/parser.cpp
#include "parser.h"
#include "error.h"
#include "expression.h"
#include "program.h"

namespace taolang {

std::map<TokenType,Precedence> precedenceTable = {
	{ttQuestion,           Precedence::Conditional},
	{ttLogicalNot,         Precedence::LogicalNot},
	{ttLogicalOr,          Precedence::LogicalOr},
	{ttLogicalAnd,         Precedence::LogicalAnd},
	{ttBitAnd,             Precedence::BitwiseAnd},
	{ttBitOr,              Precedence::BitwiseOr},
	{ttBitXor,             Precedence::BitwiseXor},
	{ttBitAndNot,          Precedence::BitwiseAndNot},
	{ttEqual,              Precedence::Equality},
	{ttNotEqual,           Precedence::Equality},
	{ttGreaterThan,        Precedence::Comparison},
	{ttGreaterThanOrEqual, Precedence::Comparison},
	{ttLessThan,           Precedence::Comparison},
	{ttLessThanOrEqual,    Precedence::Comparison},
	{ttLeftShift,          Precedence::BitwiseShift},
	{ttRightShift,         Precedence::BitwiseShift},
	{ttAddition,           Precedence::Addition},
	{ttSubtraction,        Precedence::Addition},
	{ttMultiply,           Precedence::Multiplication},
	{ttDivision,           Precedence::Multiplication},
	{ttPercent,            Precedence::Multiplication},
	{ttStarStar,           Precedence::Exponentiation},
	{ttIncrement,          Precedence::PrefixIncrement},
	{ttDecrement,          Precedence::PrefixDecrement},
	{ttLeftBracket,        Precedence::Indexing},
	{ttDot,                Precedence::Indexing},
	{ttNew,                Precedence::New},
	{ttLeftParen,          Precedence::Call},
};

Program* Parser::Parse() {
    auto program = new Program();
    for(;!_follow(ttEOF);) {
        auto stmt = _parseStatement(true);
        program->_stmts.push_back(stmt);
    }
    if(_next().type != ttEOF) {
        throw SyntaxError("unexpected token");
    }
    return program;
}

Token Parser::_expect(TokenType tt) {
    auto next = _next();
    if(next.type != tt) {
        Token tk;
        tk.type = tt;
        auto exp = tk.string();
        if(tt == ttIdentifier) {
            exp ="`identifier'";
        }
        throw SyntaxError(
            "unexpected token: %s (expect: %s)",
            next.string().c_str(), tk.string().c_str()
        );
    }
    return next;
}

Token Parser::_next() {
    return _tkz->Next();
}

void Parser::_undo(Token tk) {
    _tkz->Undo(tk);
}

bool Parser::_skip(TokenType tt) {
    if(_follow(tt)) {
        _next();
        return true;
    }
    return false;
}

Token Parser::_peek() {
    return _tkz->Peek();
}

bool Parser::_follow(TokenType tt) {
    return _peek().type == tt;
}

void Parser::_enter() {

}

void Parser::_leave(bool putBack) {

}

Precedence Parser::_getPrecedence(TokenType op) {
    if(op >= ttAssign && op <= ttAndNotAssign) {
        return Precedence::Assignment;
    }
    auto iter = precedenceTable.find(op);
    if(iter != precedenceTable.cend()) {
        return iter->second;
    }
    return Precedence::_Unspecified;
}

BaseStatement* Parser::_parseStatement(bool global) {
    auto tk = _peek();

    switch(tk.type) {
    case ttLet:
        return _parseLetStatement();
    case ttFunction:
        return _parseFunctionStatement();
    case ttSemicolon:
        _next();
        return new EmptyStatement();
    default:
        break;
    }

    if (global) {
        throw SyntaxError("non-global statement");
    }

    switch(tk.type) {
    case ttReturn:
        return _parseReturnStatememt();
    case ttLeftBrace:
        return _parseBlockStatement();
    case ttFor:
        return _parseForStatement();
    case ttBreak:
        return _parseBreakStatement();
    case ttIf:
        return _parseIfStatement();
    case ttSwitch:
        //return _parseSwitchStatement();
    default:
        break;
    }

    {
        auto expr = _parseExpression(Precedence::Assignment);
        Token tk;
        bool match;
        std::tie(tk, match) = _match(ttSemicolon);
        if(match) {
            auto es = new ExpressionStatement();
            es->_expr = expr;
            return es;
        }
        _expect(ttSemicolon);
    }

    throw SyntaxError("unknown statement at line: %d", tk.line);
}

// done
LetStatement* Parser::_parseLetStatement() {
    auto let = new LetStatement();
    _expect(ttLet);
    let->_name = _expect(ttIdentifier).str;
    if(_follow(ttAssign)) {
        _next();
        let->_expr = _parseExpression(Precedence::Conditional);
    }
    _expect(ttSemicolon);
    return let;
}

// done
FunctionStatement* Parser::_parseFunctionStatement() {
    auto fn = new FunctionStatement();
    fn->_expr = _parseFunctionExpression();
    return fn;
}

// done
ReturnStatement* Parser::_parseReturnStatememt() {
    auto rtn = new ReturnStatement();
    _expect(ttReturn);
    if(!_follow(ttSemicolon)) {
        rtn->_expr = _parseExpression(Precedence::Conditional);
    }
    _expect(ttSemicolon);
    return rtn;
}

// done
BlockStatement* Parser::_parseBlockStatement() {
    auto block = new BlockStatement();
    _expect(ttLeftBrace);
    for(;;) {
        if(_follow(ttRightBrace)) {
            break;
        }
        auto stmt = _parseStatement(false);
        block->_stmts.push_back(stmt);
    }
    _expect(ttRightBrace);
    return block;
}

// done
ForStatement* Parser::_parseForStatement() {
    auto fs = new ForStatement();
    auto hasInit = false;

    _expect(ttFor);

    if(_follow(ttLet)) {
        hasInit = true;
        fs->_init = _parseLetStatement();
    } else if(_follow(ttSemicolon)) {
        hasInit = true;
        _expect(ttSemicolon);
    } else if(!_follow(ttLeftBrace)) {
        hasInit = false;
        fs->_test = _parseExpression(Precedence::Conditional);
    }

    if(hasInit) {
        // test
        if(!_follow(ttSemicolon)) {
            fs->_test = _parseExpression(Precedence::Conditional);
            _expect(ttSemicolon);
        } else {
            _next();
        }
        // incr
        if(!_follow(ttLeftBrace)) {
            fs->_incr = _parseExpression(Precedence::Assignment);
        }
    } else {
        if(!_follow(ttLeftBrace)) {
            throw SyntaxError("for needs body");
        }
    }

    _breakCount++;

    fs->_block = _parseBlockStatement();
    
    _breakCount--;

    return fs;
}

// done
BreakStatement* Parser::_parseBreakStatement() {
    _expect(ttBreak);
    _expect(ttSemicolon);
    return new BreakStatement();
}

// done
IfStatement* Parser::_parseIfStatement() {
    auto stmt = new IfStatement();
    _expect(ttIf);
    stmt->_cond = _parseExpression(Precedence::Conditional);
    stmt->_ifBlock = _parseBlockStatement();
    if(_follow(ttElse)) {
        _next();
        switch(_peek().type) {
        case ttIf:
            stmt->_elseBlock = _parseIfStatement();
            break;
        case ttLeftBrace:
            stmt->_elseBlock = _parseBlockStatement();
            break;
        default:
            throw SyntaxError("else expect if or block to follow");
        }
    }
    return stmt;
}

IExpression* Parser::_parseExpression(Precedence prec) {
    IExpression* left = nullptr;
    auto peek = _peek();
    if(prec <= Precedence::UnaryPlus) {
        switch(peek.type) {
        case ttLogicalNot:
        case ttBitXor:
        case ttAddition:
        case ttSubtraction: {
            _next();
            auto right = _parseExpression(Precedence::UnaryPlus);
            left = new UnaryExpression(peek.type, right);
            break;
        }
        default:
            break;
        }
    }
    if(prec <= Precedence::PrefixIncrement) {
        switch(peek.type) {
        case ttIncrement:
        case ttDecrement: {
            _next();
            auto right = _parseExpression(Precedence::PrefixIncrement);
            left = new IncrementExpression(peek.type, true, right);
            break;
        }
        default:
            break;
        }
    }
    if(prec <= Precedence::New) {
        if(peek.type == ttNew) {
            left = _parseNewExpression();
        }
    }
    if(left == nullptr) {
        left = _parsePrimaryExpression();
    }

    while((1)) {
        auto op = _next();
        auto nextPrec = _getPrecedence(op.type);
        if(nextPrec == Precedence(0) || nextPrec < prec) {
            _undo(op);
            break;
        }

        if(op.type >= ttAssign && op.type < ttQuestion) {
            return _parseAssignmentExpression(left, op.type);
        }

        switch(op.type) {
        case ttQuestion:
            left = _parseTernaryExpression(left);
            continue;
        case ttIncrement:
        case ttDecrement:
            left = new IncrementExpression(op.type, false, left);
            continue;
        case ttLeftParen:
            _undo(op);
            left = _parseCallExpression(left);
            continue;
        case ttLeftBracket:
        case ttDot:
            _undo(op);
            left = _parseIndexExpression(left);
            continue;
        default:
            break;
        }

        IExpression* right = nullptr;

        switch(op.type) {
        case ttLogicalOr:
            right = _parseExpression(Precedence::LogicalAnd);
            break;
        case ttLogicalAnd:
            right = _parseExpression(Precedence::BitwiseAnd);
            break;
        case ttBitAnd:
        case ttBitOr:
        case ttBitXor:
        case ttBitAndNot:
            right = _parseExpression(Precedence::Equality);
            break;
        case ttEqual:
        case ttNotEqual:
            right = _parseExpression(Precedence::Comparison);
            break;
        case ttGreaterThan:
        case ttGreaterThanOrEqual:
        case ttLessThan:
        case ttLessThanOrEqual:
            right = _parseExpression(Precedence::BitwiseShift);
            break;
        case ttLeftShift:
        case ttRightShift:
            right = _parseExpression(Precedence::Addition);
            break;
        case ttAddition:
        case ttSubtraction:
            right = _parseExpression(Precedence::Multiplication);
            break;
        case ttMultiply:
        case ttDivision:
        case ttPercent:
            right = _parseExpression(Precedence::Exponentiation);
            break;
        case ttStarStar:
            right = _parseExpression(Precedence::Indexing);
            break;
        default:
            break;
        }

        if(right != nullptr) {
            left = new BinaryExpression(left, op.type, right);
            continue;
        }

        throw SyntaxError("unknown operator");
    }

    return left;
}

IExpression* Parser::_parsePrimaryExpression() {
    IExpression* expr;
    auto next = _next();
    switch(next.type) {
    case ttNil:
        expr = Value::fromNil();
        break;
    case ttBoolean:
        expr = Value::fromBoolean(next.str == "true");
        break;
    case ttNumber:
        expr = Value::fromNumber(next.num);
        break;
    case ttString:
        expr = Value::fromString(next.str);
        break;
    case ttLeftParen:
        _undo(next);
        if(auto lambda = _tryParseLambdaExpression(false)) {
            return lambda;
        }
        _next();
        expr = _parseExpression(Precedence::Conditional);
        _expect(ttRightParen);
        break;
    case ttIdentifier:
        if(_follow(ttLambda)) {
            _undo(next);
            return _tryParseLambdaExpression(true);
        }
        expr = Value::fromVariable(next.str);
        break;
    case ttFunction:
        _undo(next);
        expr = _parseFunctionExpression();
        break;
    case ttLeftBrace:
        _undo(next);
        expr = _parseObjectExpression();
        break;
    case ttLeftBracket:
        _undo(next);
        expr = _parseArrayExpression();
        break;
    default:
        throw SyntaxError("unexpected token");
    }

    return expr;
}

TernaryExpression* Parser::_parseTernaryExpression(IExpression* cond) {
    IExpression *left;
    IExpression *right;

    left = _parseExpression(Precedence::Conditional);
    _expect(ttColon);
    right = _parseExpression(Precedence::Conditional);

/*
    static const char* err = "nested `?:' is not allowed";
    if(left->type == ExprType::Ternary) {
        throw SyntaxError(err);
    }
    if(right->type == ExprType::Ternary) {
        throw SyntaxError(err);
    }
    */

    auto expr = new TernaryExpression();
    expr->left = left;
    expr->cond = cond;
    expr->right = right;

    return expr;
}

AssignmentExpression* Parser::_parseAssignmentExpression(IExpression* left, TokenType op) {
    auto expr = new AssignmentExpression();
    expr->_left = left;

	// ttQuestion: disable continuous assignment style
    expr->_expr = _parseExpression(Precedence::Conditional);

	if(op == ttAssign) {
		return expr;
	}

    TokenType binOp;

	switch(op) {
	case ttStarStarAssign:
		binOp = ttStarStar;
        break;
	case ttStarAssign:
		binOp = ttMultiply;
        break;
	case ttDivideAssign:
		binOp = ttDivision;
        break;
	case ttPercentAssign:
		binOp = ttPercent;
        break;
	case ttPlusAssign:
		binOp = ttAddition;
        break;
	case ttMinusAssign:
		binOp = ttSubtraction;
        break;
	case ttLeftShiftAssign:
		binOp = ttLeftShift;
        break;
	case ttRightShiftAssign:
		binOp = ttRightShift;
        break;
	case ttAndAssign:
		binOp = ttBitAnd;
        break;
	case ttOrAssign:
		binOp = ttBitOr;
        break;
	case ttXorAssign:
		binOp = ttBitXor;
        break;
	case ttAndNotAssign:
		binOp = ttBitAndNot;
        break;
	default:
        throw Error("won't go here");
	}

    auto bin = new BinaryExpression();
    bin->_left = expr->_left;
    bin->_op = binOp;
    bin->_right = expr->_expr;

    expr->_expr = bin;
    return expr;
}

NewExpression* Parser::_parseNewExpression() {
    auto expr = new NewExpression();
    _expect(ttNew);
    expr->_name = _expect(ttIdentifier).str;
    _expect(ttLeftParen);
    if(!_follow(ttRightParen)) {
        for (;;) {
            auto arg = _parseExpression(Precedence::Conditional);
            expr->_args.Put(arg);
            auto sep = _next();
            if(sep.type == ttComma) {
                continue;
            } else if(sep.type == ttRightParen) {
                _undo(sep);
                break;
            } else {
                throw SyntaxError("unexpected token");
            }
        }
    }
    _expect(ttRightParen);
    return expr;
}

FunctionExpression* Parser::_tryParseLambdaExpression(bool must) {
    _enter();
}

IndexExpression* Parser::_parseIndexExpression(IExpression* left) {
    auto ie = new IndexExpression();
    ie->_indexable = left;
    auto next = _next();
    if(next.type == ttDot) {
        auto key = _next();
        if(key.type == ttIdentifier) {
            ie->_key = Value::fromString(key.str);
            return ie;
        }
        throw SyntaxError("unexpected token: %s", key.string().c_str());
    } else if(next.type == ttLeftBracket) {
        auto key = _parseExpression(Precedence::Conditional);
        _expect(ttRightBracket);
        ie->_key = key;
        return ie;
    } else {
        throw Error("won't go here");
    }
}

CallExpression* Parser::_parseCallExpression(IExpression* left) {
    auto ce = new CallExpression();
    ce->_callable = left;
    _expect(ttLeftParen);
    if(!_follow(ttRightParen)) {
        for(;;) {
            auto arg = _parseExpression(Precedence::Conditional);
            ce->_args.Put(arg);
            auto sep = _next();
            if(sep.type == ttComma) {
                continue;
            } else if(sep.type == ttRightParen) {
                _undo(sep);
                break;
            } else {
                SyntaxError(
                    "unexpected token: %s",
                    Token(sep).string().c_str()
                );
            }
        }
    }
    _expect(ttRightParen);
    return ce;
}

FunctionExpression* Parser::_parseFunctionExpression() {
    auto fe = new FunctionExpression();

    _expect(ttFunction);

    if(_follow(ttIdentifier)) {
        fe->_name = _next().str;
    }

    _expect(ttLeftParen);
    if(!_follow(ttRightParen)) {
        for(;;) {
            auto name = _expect(ttIdentifier).str;
            fe->_params.Put(name);
            auto sep = _next();
            if(sep.type == ttComma) {
                continue;
            } else if(sep.type == ttRightParen) {
                _undo(sep);
                break;
            } else {
                throw SyntaxError(
                    "unexpected token: %s",
                    Token(sep).string().c_str()
                );
            }
        }
    }
    _expect(ttRightParen);

    if(!_follow(ttLeftBrace)) {
        throw SyntaxError("function needs a body");
    }

    auto savedBreakCount = _breakCount;
    _breakCount = 0;
    fe->_body = _parseBlockStatement();
    _breakCount = savedBreakCount;
    return fe;
}

ObjectExpression* Parser::_parseObjectExpression() {
    auto obj = new ObjectExpression();
    _expect(ttLeftBrace);
    for(;;) {
        if(!_follow(ttRightBrace)) {
            break;
        }

        std::string key;
        IExpression* val;

        auto next = _next();
        switch(next.type) {
        case ttString:
        case ttIdentifier:
            key = next.str;
            break;
        default:
            throw TypeError("unsupported key type");
        }

        _expect(ttColon);

        val = _parseExpression(Precedence::Conditional);
        auto iter = obj->_props.find(key);
        if(iter == obj->_props.cend()) {
            obj->_props[key] = val;
        } else {
            throw SyntaxError("duplicate key");
        }

        _skip(ttComma);
        if(_follow(ttRightBrace)) {
            break;
        }
    }

    _expect(ttRightBrace);
    return obj;
}

ArrayExpression* Parser::_parseArrayExpression() {
    auto arr = new ArrayExpression();
    _expect(ttLeftBracket);
    for(;;) {
        if(_follow(ttRightBracket)) {
            break;
        }
        auto elem = _parseExpression(Precedence::Conditional);
        arr->_elems.Put(elem);
        _skip(ttComma);
        if(_follow(ttRightBracket)) {
            break;
        }
    }
    _expect(ttRightBracket);
    return arr;
}

}
